server/modifiers.py

```diff
@@ -0,0 +1,202 @@
+        for num in line_subset:
+            self.modified_contents[num] = self.modified_contents[num].replace(":", "")
+
+        return self
+
+    def change_keyword(self) -> Self:
+        """A code modifier that causes a SyntaxError.
+
+        This will change any of the python keywords to "kappa".
+
+        Returns:
+            The modifier instance.
+        """
+        python_keywords = keyword.kwlist
+
+        number_keyword_pairs = []
+        for num, line in enumerate(self.file_contents):
+            if any(key in line for key in python_keywords):
+                number_keyword_pairs.extend([(num, key) for key in python_keywords if key in line])
+
+        line_subset = random.sample(number_keyword_pairs, min(self.difficulty, len(number_keyword_pairs)))
+        for num, key in line_subset:
+            self.modified_contents[num] = self.modified_contents[num].replace(key, random.choice(STATEMENTS))
+
+        return self
+
+    def comment(self) -> Self:
+        """A code modifier that could raise an error.
+
+        This will comment out a line of code.
+
+        Returns:
+            The modifier instance.
+        """
+        line_numbers = []
+        for num, line in enumerate(self.file_contents):
+            if line != "\n":
+                line_numbers.append(num)
+
+        line_subset = random.sample(line_numbers, min(self.difficulty, len(line_numbers)))
+        for num in line_subset:
+            self.modified_contents[num] = f"# {self.modified_contents[num]}"
+
+        return self
+
+    def change_function_call_name(self) -> Self:
+        """A code modifier that causes a NameError.
+
+        Where a function is called, the name of that function will be changed to `cj9_kappa`.
+
+        Returns:
+            The modifier instance.
+        """
+        function_names = []
+        for num, line in enumerate(self.file_contents):
+            match = STARTSWITH_DEF_REGEX.match(line)
+
+            if match:
+                func_name = match.groups()[1].split("(")[0]
+
+                # Don't include dunder methods
+                if func_name.startswith("__"):
+                    continue
+
+                # If the method is a property, don't use it as it's not callable
+                if self.file_contents[num - 1] == f"{FOUR_SPACES}@property\n":
+                    continue
+
+                function_names.append((num, func_name))
+
+        line_subset = random.sample(function_names, min(self.difficulty, len(function_names)))
+        for num, line in enumerate(self.file_contents):
+            for def_num, func_name in line_subset:
+                if num == def_num:
+                    continue
+
+                func_match = re.match(rf".*\.?({func_name}\().*", line)
+
+                if func_match:
+                    self.modified_contents[num] = self.modified_contents[num].replace(func_name, "cj9_kappa")
+
+        return self
+
+    def insert_empty_statements(self) -> Self:
```

Comment:
`insert_empty_statement` inserts multiple newline characters in the line, so I was wondering if they should be returned as multiple lines that were changed, instead of one single line.

If so, the `\n` at the end of every line might not be needed, our editor could just assume that there is a new line between consecutive list elements, which could make the code look a bit cleaner.

For now, I don't think that's an important change, but we might consider it in the future.

Reply:
They could be returned as multiple lines, it would just mean that the modified code would be even more buggy, for example a perfectly good line of code would be replaced by a `\n`. I could remove the `\n` from `self.modified_contents[random_position] = f"{self.modified_contents[random_position]}\n{statement}"` just before the statement at the end as I only included it to look better, essentially. If that was removed, then this would only affect one line and not two.

Reply:
> If so, the `\n` at the end of every line might not be needed, our editor could just assume that there is a new line between consecutive list elements, which could make the code look a bit cleaner.

I think the newlines need to be included when receiving the data from the client at least. For some reason, the diff doesn't get created without newlines, or just generally without something changing between the input and the output.

For example, try removing the `\n` from the `test_lines` and also the `[:-1]` from the instance attributes in the \_\_init\_\_ and you will always get `[(0, "")]` as the return value. If you can somehow find a way to make this work without relying on any of the modifications from receiving the data, then feel free to PR that into this branch.